components/render/atlastexture.cpp

```diff
@@ -0,0 +1,148 @@
+/* Stores many small textures into a large texture (or array of textures)
+ * to allow batched drawing commands that increase performance.
+ * Currently implemented as an array of 2D textures, would be simpler with much
+ * better performance if a proper 2D texture array (GL_TEXTURE_2D_ARRAY) was
+ * used. Author couldn't quite get it to work, parts of the Nuklear GUI would
+ * be missing if the array texture was allocated with > 1 depth (layers)... */
+
+#include "atlastexture.h"
+#include "../../extern/RectangleBinPack/SkylineBinPack.h"
+
+#include <boost/make_unique.hpp>
+#include <misc/assert.h>
+
+namespace Render
+{
+    AtlasTexture::AtlasTexture()
+    {
+        GLint maxTextureSize;
+        GLint maxTextures = 8; // Hardcoded in fragment shader, GL 3.x minimum is 16.
+        glGetIntegerv(GL_MAX_TEXTURE_SIZE, &maxTextureSize);
+
+        mTextureWidth = maxTextureSize;
+        mTextureHeight = maxTextureSize;
+
+        // Limit number of textures to a reasonable level (measured from testing).
+        // Note: Increasing this has a severe impact on performance.
+        GLint estimatedRequiredTextures = (1uLL << 29) / (mTextureWidth * mTextureHeight);
+        mTextureLayers = std::min(estimatedRequiredTextures, maxTextures);
```

Comment:
I'm getting maxTextureSize = 32768 on my GTX 1050. This makes mTextureWidth * mTextureHeight be equal to 1073741824, which makes the division in estimatedRequiredTextures be equal to 0, making mTextureLayers be equal to 0. This leads to a assertion failure (release_assert(layer < mTextureLayers)).